lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -450,4 +464,192 @@ def _env_manager_for(self, type: ServiceType) -> _EnvManager:
+D4Q2ZBsqlN+5aW2q0IUmO+m0GhcdaDv8U7RVto1cWWPr50HhiCi7Yvei1qZiD9jq
+57oYZVqTUNCTPxi6NeTOdEc+YqNynWNArx4PHh38LT0bqKtlZCGHNfoAJLPVYhbB
+b2AHj9edYtHU9AAFSIy+HstET6P0UDxy02IeyE2yxoUBqdlXyv6FL44E+wARAQAB
+zRxMYXVuY2hwYWQgUFBBIGZvciBVYnVudHUgSFBDwsGOBBMBCgA4FiEErocSHcPk
+oLD4H/Aj9tDF1ca+s3sFAmTuZb8CGwMFCwkIBwIGFQoJCAsCBBYCAwECHgECF4AA
+CgkQ9tDF1ca+s3sz3w//RNawsgydrutcbKf0yphDhzWS53wgfrs2KF1KgB0u/H+u
+6Kn2C6jrVM0vuY4NKpbEPCduOj21pTCepL6PoCLv++tICOLVok5wY7Zn3WQFq0js
+Iy1wO5t3kA1cTD/05v/qQVBGZ2j4DsJo33iMcQS5AjHvSr0nu7XSvDDEE3cQE55D
+87vL7lgGjuTOikPh5FpCoS1gpemBfwm2Lbm4P8vGOA4/witRjGgfC1fv1idUnZLM
+TbGrDlhVie8pX2kgB6yTYbJ3P3kpC1ZPpXSRWO/cQ8xoYpLBTXOOtqwZZUnxyzHh
+gM+hv42vPTOnCo+apD97/VArsp59pDqEVoAtMTk72fdBqR+BB77g2hBkKESgQIEq
+EiE1/TOISioMkE0AuUdaJ2ebyQXugSHHuBaqbEC47v8t5DVN5Qr9OriuzCuSDNFn
+6SBHpahN9ZNi9w0A/Yh1+lFfpkVw2t04Q2LNuupqOpW+h3/62AeUqjUIAIrmfeML
+IDRE2VdquYdIXKuhNvfpJYGdyvx/wAbiAeBWg0uPSepwTfTG59VPQmj0FtalkMnN
+ya2212K5q68O5eXOfCnGeMvqIXxqzpdukxSZnLkgk40uFJnJVESd/CxHquqHPUDE
+fy6i2AnB3kUI27D4HY2YSlXLSRbjiSxTfVwNCzDsIh7Czefsm6ITK2+cVWs0hNQ=
+=cs1s
+-----END PGP PUBLIC KEY BLOCK-----
+"""
+
+
+class AptManager(SlurmOpsManager):
+    """Slurm ops manager that uses apt as its package manager."""
+
+    def install(self) -> None:
+        """Install Slurm using the `slurm` snap."""
+        slurm_wlm = apt.DebianRepository(
+            enabled=True,
+            repotype="deb",
+            uri="https://ppa.launchpadcontent.net/ubuntu-hpc/slurm-wlm-23.02/ubuntu",
+            release=distro.codename(),
+            groups=["main"],
+        )
+        slurm_wlm.import_key(UBUNTU_HPC_SLURM_WLM_KEY)
+
+        experimental = apt.DebianRepository(
+            enabled=True,
+            repotype="deb",
+            uri="https://ppa.launchpadcontent.net/ubuntu-hpc/experimental/ubuntu",
+            release=distro.codename(),
+            groups=["main"],
+        )
+        experimental.import_key(UBUNTU_HPC_EXPERIMENTAL_KEY)
+
+        repositories = apt.RepositoryMapping()
+        repositories.add(slurm_wlm)
+        repositories.add(experimental)
+
+        apt.update()
+        for package in ["slurm-wlm", "mungectl", "prometheus-slurm-exporter"]:
+            try:
+                apt.add_package(package)
+            except apt.PackageNotFoundError as e:
+                raise SlurmOpsError(f"package {package} not found. reason: {e}")
+            except apt.PackageError as e:
+                raise SlurmOpsError(f"failed to install package {package}. reason: {e}")
+
+        override = Path("/etc/systemd/system/slurmd.service.d/10-slurmd-conf-server.conf")
+        override.parent.mkdir(exist_ok=True, parents=True)
+        override.write_text(
+            textwrap.dedent(
+                """
+                [Service]
+                ExecStart=/usr/bin/sh -c "/usr/sbin/slurmd -D -s $${SLURMD_CONFIG_SERVER:+--conf-server $$SLURMD_CONFIG_SERVER} $$SLURMD_OPTIONS"
+                """
+            )
+        )
```

Comment:
This looks good, but for the relevant `slurmd` bits, we should ensure that these methods don't conflict with what's been defined here for the slurmd service management utility in the slurmd charm:

https://github.com/charmed-hpc/slurm-charms/blob/246f1da25861766b0b42a020f3b40bff4bf20f32/charms/slurmd/src/utils/slurmd.py#L50-L96

Reply:
I think ensuring this should be the responsibility of the charms, and we should just give a sane default and document that the manager uses `/etc/default/slurmd` and `SLURMD_CONFIG_SERVER` to do that.

Reply:
I'll still move the env vars redefinitions to different files, but the charm would need to ensure that the `SLURMD_CONFIG_SERVER` parameter is correctly set.

Reply:
Sounds good. We'll take note of this when we further integrate `slurm_ops` into the charms.